<a href="https://colab.research.google.com/github/giedriusblazys/SDA_DS_class/blob/main/SDA_DS_Project_PUBLIC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

https://www.kaggle.com/mirbektoktogaraev/madrid-real-estate-market

| Variable Name  | Description            |
| -----------    | -----------            |
| ID             | Identifier ID          |
| title          | Title from listing     |
| subtitle       | Neighborhood and city  |
| sqmtbuilt      |  Square meter built    |
| sqmtuseful     | Square meter useful    |
| n_rooms        | Number of rooms        |
| n_bathrooms    | Number of bathrooms    |
| sqmtallotment  | Square meter allotment |
| latitude       | Latitude (Since portals hide the location mainly NAs)  |
| longitude      | Longitude (Since portals hide the location mainly NAs) |
| raw_address    | Address                |
| isexactaddress_hidden | Boolean values  |
| buy_price             | Target Value    |


2018 3,223,334

#pandas_profiling

In [ ]:
#!pip install pandas-profiling==2.7.1

In [ ]:
#!pip install pandas_profiling

In [ ]:
import pandas_profiling
'''
pip install pandas_profiling
import pandas_profiling
import pandas as pd
df = pdf.read_csv("x.csv")
house_price_report=pandas_profiling.ProfileReport(df)
house_price_report.to_file('house_report.html')
'''

'\npip install pandas_profiling\nimport pandas_profiling\nimport pandas as pd\ndf = pdf.read_csv("x.csv")\nhouse_price_report=pandas_profiling.ProfileReport(df)\nhouse_price_report.to_file(\'house_report.html\')\n'

In [ ]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

# Data set: houses_Madrid.csv

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/linavait/2021-11-20/master/namu_darbai/houses_Madrid.csv') 
                 #index_col=[0])#index_col='id')
full_df = df
df['operation'].value_counts()

sale    21742
Name: operation, dtype: int64

In [ ]:
#house_price_report = pandas_profiling.ProfileReport(df)
#house_price_report.to_file('house_report.html')

In [ ]:
df['buy_price_per_sq_mt_built'] = df.buy_price/df.sq_mt_built
df['buy_price_per_sq_mt_built'].median()

3732.8638497652582

#Util functions

In [ ]:
def percent_from_total(x):#fraction
  return str(round(x/len(df)*100, 1))+' %'

percent_from_total(8_000) # Test

'36.8 %'

In [ ]:
print(type(3))
print(type(3.5))
print(type('3.5'))

<class 'int'>
<class 'float'>
<class 'str'>


# Address to coordinates conversion

In [ ]:
df.head(3)
i = 0
for title in df['title']:
  if "calle" not in title: 
    # calle yra gatvė. Kartais nėra žodžio gatvė, bet yra <gatvės pavadinimas>, <gatvės namo numeris>
    #print(title)
    i += 1
#print(i)
'''
https://developers.google.com/maps/documentation/geocoding/requests-geocoding
https://maps.googleapis.com/maps/api/geocode/json?address=Savanori%C5%B3%20pr.%201,%20Kaunas
Test is here https://developers.google.com/maps/documentation/geocoding/overview#Geocoding
20k reqs (kurie šiaip kainuotų 200$) yra nemokami https://mapsplatform.google.com/pricing/
https://github.com/googlemaps/google-maps-services-python
'''
''

''

In [ ]:
# df.dtypes

# df.describe()

In [ ]:
df.describe()

Unnamed: 0  ...  buy_price_per_sq_mt_built
count             21742.00  ...                   21616.00
mean              10870.50  ...                    4026.97
std                6276.52  ...                    1910.02
min                   0.00  ...                     447.37
25%                5435.25  ...                    2554.35
50%               10870.50  ...                    3732.86
75%               16305.75  ...                    5000.00
max               21741.00  ...                   18888.89

[8 rows x 24 columns]

# Darbų planas (2022-02-13)

In [ ]:
'''
Madrid real estate data:
  21742 rows
  58 columns

9:30-12:00
  1. Išsirinkome įrankius projektui (colab + anydesk + messenger + Google Docs for slides) + mandatory Zoom.
  2. (2,5 val.) apsirašėme 56 features.
12:00 išsirinkome 6 feature'us pagal koreliaciją 15 int tipo featurų (int'ų tipo feature'ų yra 27 % (<-15/56)):
features = [
  'sq_mt_built', # 65- 770m2, NaN(126), reikšmių(21616)
  'sq_mt_useful',
  'n_rooms',
  'n_bathrooms',
  'buy_price_by_area',
  'has_parking'
]
Lina: SWOT
Alina: LinearRegression
Kęstutis: DecitionTrees

13:33  dėstytojas atėjo visi užsimutinę mikrofonus, pamatė, kad jau skaidres dėliojame... ir sako tai gal aš jau einu į kitą... ir išėjo nespėjus...



1. Išsirinkti features, kad nebūtų daug nan reikštmių.
? 2. Konvertuoti nan reikšmes į false, jei nėra false reikšmių
3. atsirinkti tik tas eilutes kur nėra nan reikšmių.
4. paleisti: linear regression, random forest (decition trees)

13:15- skaidres daryti
13:35-13:50 Kęstutis valgo
14:00 pristatinėjam
'''

"\nMadrid real estate data:\n  21742 rows\n  58 columns\n\n9:30-12:00\n  1. Išsirinkome įrankius projektui (colab + anydesk + messenger + Google Docs for slides) + mandatory Zoom.\n  2. (2,5 val.) apsirašėme 56 features.\n12:00 išsirinkome 6 feature'us pagal koreliaciją 15 int tipo featurų (int'ų tipo feature'ų yra 27 % (<-15/56)):\nfeatures = [\n  'sq_mt_built', # 65- 770m2, NaN(126), reikšmių(21616)\n  'sq_mt_useful',\n  'n_rooms',\n  'n_bathrooms',\n  'buy_price_by_area',\n  'has_parking'\n]\nLina: SWOT\nAlina: LinearRegression\nKęstutis: DecitionTrees\n\n13:33  dėstytojas atėjo visi užsimutinę mikrofonus, pamatė, kad jau skaidres dėliojame... ir sako tai gal aš jau einu į kitą... ir išėjo nespėjus...\n\n\n\n1. Išsirinkti features, kad nebūtų daug nan reikštmių.\n? 2. Konvertuoti nan reikšmes į false, jei nėra false reikšmių\n3. atsirinkti tik tas eilutes kur nėra nan reikšmių.\n4. paleisti: linear regression, random forest (decition trees)\n\n13:15- skaidres daryti\n13:35-13:50 Kęs

# Features

## Code for a feature description (v1)

In [ ]:
feature = 'has_individual_heating'
BOLD, END = '\033[1m', '\033[0m'
display(df[feature].head())
if df[feature].dtype in ['int64','float64']:#dtype: object
  print(BOLD + str(df[feature].min())+ '-'+ str(df[feature].max()) + END)
else:
  print('No min-max because '+str(df[feature].dtype))
if True in df[feature].isna().value_counts():
  print(BOLD + 'NaN('+str(df[feature].isna().value_counts()[True])+')' + END)
print(BOLD + 'value_counts:' + END)
display(df[feature].value_counts())

0     NaN
1     NaN
2    True
3     NaN
4     NaN
Name: has_individual_heating, dtype: object

No min-max because object
NaN(8134)
value_counts:


True     9516
False    4092
Name: has_individual_heating, dtype: int64

## Description of all features

In [ ]:
'''
| Variable Name  | Description            | Example
| -----------    | -----------            | -----------
| ID             | Identifier ID          |
| title          | Title from listing     | Piso en venta en calle de Godella, 64 -> Butas parduodamas adresu Calle de Godella, 64
| subtitle       | Neighborhood and city  | San Cristóbal, Madrid
| sq_mt_built    | Square meter built     | 65- 770m2, NaN(126), kai su reikšmėm - 21616  
| sq_mt_useful   | Square meter useful    | 16 - 778m2, NaN(13514), kai su reikšmėm - 8228 
| n_rooms        | Number of rooms        | 1-24
| n_bathrooms    | Number of bathrooms    | 1-16
| n_floors       |                        | NaN(20305), 1-7. Išskyrus reikšmę 6.
| sq_mt_allotment| Square meter allotment | 1-382, NaN(20310), kai su reikšmėm - 1432    
| latitude       | Latitude (Since portals hide the location mainly NAs)  | 21742- NaN, NĖRA reikšmių
| longitude      | Longitude (Since portals hide the location mainly NAs) | 21742- NaN, NĖRA reikšmių
| raw_address    | Address                | NaN(5465), kai su rekšmėm (16277)
| is_exact_address_hidden | Boolean values| True-15002; False-6740 -> NaN(15002), kai su rekšmėm (6740)
| street_name    | street name            | NaN(5905)  , kai su reikšmėm - 15837
| street_number  | street number          | 23 - A, NaN(15442), kai su reikšmėm - 6300
-| portal         |                        | NaN(21742), NĖRA reikšmių
| floor          |                        | NaN(2607), kai su reikšmėm  - 19135
  1                       4440
  2                       3546
  3                       3001
  4                       2323
  Bajo                    2144 #Apatinis aukštas arba namo aukštas, esantis gatvės lygyje.
  5                       1310
  6                        913
  7                        556
  8                        326
  Entreplanta exterior     236 #Floor plan of a house or building that is built by removing part of the height from the lower or upper floor.
  9                        181
  Semi-sótano exterior      55 | pusrūsio eksterjeras
  #sótano=rūsys
  <...>
| is_floor_under |                        | False-18033; True-2539 -> NAN(1170), kai su reikšmėm - 20572
-| door           |                        | NaN(21742), NĖRA reikšmių


'''
#df['n_rooms'].head()
#.isna() 
#21742 
#df['sq_mt_built'].value_counts()

'''
| neighborhood_id | | #126 Neighborhood 23: Malasaña-Universidad (5196.25 €/m2) - District 4: Centro                485           
  Tas pats mikrorajonas/District 2
  Neighborhood 9: Campo de las Naciones-Corralejos (3417.44 €/m2) - District 2: Barajas      9
  Neighborhood 10: Casco Histórico de Barajas (3100.87 €/m2) - District 2: Barajas           6
  Neighborhood 11: Timón (2879.88 €/m2) - District 2: Barajas                                6
| operation | | sale - #21742
| rent_price| | int visi yra: -34590276-18421 #TODO: clean. Dig deeper
| rent_price_by_area | | NĖRA
| is_rent_price_known | | NĖRA
| buy_price | | 36000 - 8800000
| buy_price_by_area i| | 447-18889. #5643 of uniq. Paplitimas: 5000    143, 3000     99, ...
| is_buy_price_known | | visi True NEREIKALINGA
| house_type_id | | NaN(391)
  HouseType 1: Pisos            17705
  HouseType 2: Casa o chalet     1938 # namas ar namelis| namas ar kotedžas
  HouseType 5: Áticos            1032 # mansardiniai namai. Gal namai su mansardom?
  HouseType 4: Dúplex             676 # Dvibutis namas, bendra siena.
| is_renewal_needed | | False    17747, True      3995
| is_new_development | | NaN(992). False    19281, True      1469
| built_year | | NaN(11742)
| has_central_heating | | NaN(8134). False    9516, True     4092
-| has_individual_heating o| NaN(8134): True     9516, False    4092 //nes vienas kitą paneigia has_central_heating
| are_pets_allowed | | **NĖRA** reikšmių
| has_ac | | NaN(10531), True    11211, False reikšmių nėra!!!
| has_fitted_wardrobes | | NaN(8343), True    13399, False reikšmių nėra!!!
| has_lift | | NaN(2386), True     14895, False     4461
| is_exterior | | NaN(3043), True     16922, False     1777
'''


'''
| has_garden      | Garden                 | NaN (20186), kai su reikšmėm 1556, True - 1556 |
| has_pool        | Pool                   | NaN (16571), kai su reikšmėm 5171, True - 5171 |
| has_terrace     | Terrace                | NaN (12194), kai su reikšmėm 9548, True - 9548 |
| has_balcony     | Balcony                | NaN (18421), kai su reikšmėm 3321, True - 3321 |
| has_storage_room| Storage room           | NaN (14044), kai su reikšmėm 7698, True - 7698 |
| is_furnished    | Is furnished           | NaN (21742)                       |
| is_kitchen_equipped| Kitchen is equipped | NaN (21742)                       |
| is_accessible   | Accessible             | NaN (17668), kai su reikšmėm 4074, True - 4074 |
| has_green_zones | Green zones            | NaN (17685), kai su reikšmėm 4057, True - 4057 |
| energy_certificate | Energy certificate  | en trámite(10937), no indicado (3669), E(2701), D(1121), G(894), F(674), A(618), C(580), B(452), inmueble exento(96)|
| has_parking     | Parking                | True - 7719, False - 14023                     |
| has_private_parking | Private parking    | NaN (21742)                       |
| has_public_parking  | Public parking     | NaN (21742)                       |
| is_parking_included_in_price | Parking is included in price | NaN (14023), kai su reikšmėm 7719, True - 7115, False - 604 |
| parking_price   | Parking price          | 0-30000, NaN (14023), kai su reikšmėm 7719|
| is_orientation_north | North orientation | NaN (10384), kai su reikšmėm 10384, True - 2791, False - 8567|
| is_orientation_west  | West orientation  | NaN (10384), kai su reikšmėm 10384, True - 4071, False - 7287|
| is_orientation_south | South orientation | NaN (10384), kai su reikšmėm 10384, True - 5823, False - 5535|
| is_orientation_east  | East orientation  | NaN (10384), kai su reikšmėm 10384, True - 5056, False - 6302|
'''
print()

## All features

In [ ]:
features = [#Lina
            'id',                          
            'title',                       
            'subtitle',                    
            'sq_mt_built',                 
            'sq_mt_useful',                
            'n_rooms',                     
            'n_bathrooms',                 
            'n_floors',                    
            'sq_mt_allotment',             
            'latitude',                    
            'longitude',                   
            'raw_address',                 
            'is_exact_address_hidden',     
            'street_name',                 
            'street_number',               
            'portal',                      
            'floor',                       
            'is_floor_under',              
            'door',           

            #Kęstutis             
            'neighborhood_id',            
            'operation',                   
            'rent_price',                  
            'rent_price_by_area',          
            'is_rent_price_known',         
            'buy_price',                   
            'buy_price_by_area',           
            'is_buy_price_known',          
            'house_type_id',               
            'is_renewal_needed',           
            'is_new_development',          
            'built_year',                  
            'has_central_heating',         
            'has_individual_heating',      
            'are_pets_allowed',            
            'has_ac',                      
            'has_fitted_wardrobes',        
            'has_lift',                    
            'is_exterior',     

            #Alina
            'has_garden',                  
            'has_pool',                    
            'has_terrace',                 
            'has_balcony',                 
            'has_storage_room',            
            'is_furnished',                
            'is_kitchen_equipped',         
            'is_accessible',               
            'has_green_zones',             
            'energy_certificate',          
            'has_parking',                 
            'has_private_parking',         
            'has_public_parking',          
            'is_parking_included_in_price',
            'parking_price',               
            'is_orientation_north',        
            'is_orientation_west',       
            'is_orientation_south',        
            'is_orientation_east'
]

### Features correlation

In [ ]:
def corr_plot(mat, title='My Title', color='cividis'):
    plt.subplots(figsize=(14, 11))
    plt.title(title,fontsize=22)

    sns.heatmap(mat,linewidths=0.2,vmax=0.7,square=True,cmap=color, 
                linecolor='w',annot=True,annot_kws={'size':9},cbar_kws={'shrink': .5})

In [ ]:
corr_matrix = df[features].corr().round(3)
# corr_matrix

In [ ]:
corr_plot(corr_matrix, "Correlation matrix")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  


## Reduced features (-12 features with no values)

In [ ]:
features = [
            'id',                          
            'title',                       
            'subtitle',                    
            'sq_mt_built',                 
            'sq_mt_useful',                
            'n_rooms',                     
            'n_bathrooms',                 
            'n_floors',                    
            'sq_mt_allotment',                            
            'raw_address',                 
            'is_exact_address_hidden',     
            'street_name',                 
            'street_number',                                    
            'floor',                       
            'is_floor_under',                                     
            'neighborhood_id',            
            'operation',                   
            'rent_price',                          
            'buy_price',                   
            'buy_price_by_area',                  
            'house_type_id',               
            'is_renewal_needed',           
            'is_new_development',          
            'built_year',                  
            'has_central_heating',         
            'has_individual_heating',                 
            'has_ac',                      
            'has_fitted_wardrobes',        
            'has_lift',                    
            'is_exterior',                 
            'has_garden',                  
            'has_pool',                    
            'has_terrace',                 
            'has_balcony',                 
            'has_storage_room',                  
            'is_accessible',               
            'has_green_zones',             
            'energy_certificate',          
            'has_parking',                          
            'is_parking_included_in_price',
            'parking_price',               
            'is_orientation_north',        
            'is_orientation_west',       
            'is_orientation_south',        
            'is_orientation_east'
]

### Reduced features correlation (-12 features with no values)

In [ ]:
corr_matrix = df[features].corr().round(3)
# corr_matrix

In [ ]:
corr_plot(corr_matrix, "Correlation matrix")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  


## Code for a feature description (v2)

### dtypes

In [ ]:
dtypes = ['bool', 'int64', 'float64', 'object']
#dtypes_to_char = {'bool': 'b', 'int64':'i', 'float64':'f', 'object':'o'}
numbers = ['int64', 'float64']
i = 1
for col in df.columns:
  #if df[col].dtype in numbers:
    #print(col.ljust(28), i, str(df[col].dtype)[0])
  print(col.ljust(28), i, df[col].dtype)
  i += 1
##22 featurus

Unnamed: 0                   1 int64
id                           2 int64
title                        3 object
subtitle                     4 object
sq_mt_built                  5 float64
sq_mt_useful                 6 float64
n_rooms                      7 int64
n_bathrooms                  8 float64
n_floors                     9 float64
sq_mt_allotment              10 float64
latitude                     11 float64
longitude                    12 float64
raw_address                  13 object
is_exact_address_hidden      14 bool
street_name                  15 object
street_number                16 object
portal                       17 float64
floor                        18 object
is_floor_under               19 object
door                         20 float64
neighborhood_id              21 object
operation                    22 object
rent_price                   23 int64
rent_price_by_area           24 float64
is_rent_price_known          25 bool
buy_price                    26 

In [ ]:
print(len(df))
len(df[(~df['sq_mt_built'].isna()) & (~df['sq_mt_useful'].isna())])
#21742
#~df['sq_mt_built'].isna()

21742


8210

### Number type features

In [ ]:
i = 1

notes_inline = {
    '' : '',
}
notes_below = {
    '' : '',
}
#TODO: mixed tipus...
for split in [True, False]:
#split = True
#if True:
  for col in df.columns:
    feature = col
    if df[col].dtype in numbers and 'id' != col:
      nan = df[feature].isna().value_counts()
      #print(split, (True in nan and nan[True] == len(df)))
      if split == (True in nan and nan[True] == len(df)):
        print(str(i).ljust(2), col.ljust(28),  str(df[col].dtype)[0], end='')
        BOLD, END = '\033[1m', '\033[0m'
        #display(df[feature].head())
        
        if True in nan:
          print(' '+BOLD + 'NaN('+str(nan[True])+', '+percent_from_total(nan[True])+')' + END, end='')
        
        if df[feature].dtype in ['int64','float64']:#dtype: object
          print(' '+BOLD + str(df[feature].min())+ '-'+ str(df[feature].max()) + END, end='')
        else:
          print('No min-max because '+str(df[feature].dtype), end='')

        if not(True in nan):
          print(' uniq('+str(len(df[feature].value_counts()))+')', end='') 

          print(' TOP3: ', end='')
          for val, cnt in df[feature].value_counts().nlargest(3).items():
            print(val, '(', cnt, '); ', sep='', end='')

          print(' BOTTOM1 ', end='')
          for val, cnt in df[feature].value_counts().nsmallest(1).items():
            print(val, '(', cnt, '); ', sep='', end='')
        #display(df[feature].value_counts())
        print()
        #percent_from_total()
        
        # NaN(126, XX%), reikšmių(21616, XX%), NUO–IKI, top 3 value_counts...
        # Perdėtas NaN(126, XX%), reikšmių(21616, XX%), NUO–IKI, top 3 value_counts...
        i += 1

1  latitude                     f NaN(21742, 100.0 %) nan-nan
2  longitude                    f NaN(21742, 100.0 %) nan-nan
3  portal                       f NaN(21742, 100.0 %) nan-nan
4  door                         f NaN(21742, 100.0 %) nan-nan
5  rent_price_by_area           f NaN(21742, 100.0 %) nan-nan
6  are_pets_allowed             f NaN(21742, 100.0 %) nan-nan
7  is_furnished                 f NaN(21742, 100.0 %) nan-nan
8  is_kitchen_equipped          f NaN(21742, 100.0 %) nan-nan
9  has_private_parking          f NaN(21742, 100.0 %) nan-nan
10 has_public_parking           f NaN(21742, 100.0 %) nan-nan
11 Unnamed: 0                   i 0-21741 uniq(21742) TOP3: 0(1); 14491(1); 14499(1);  BOTTOM1 0(1); 
12 sq_mt_built                  f NaN(126, 0.6 %) 13.0-999.0
13 sq_mt_useful                 f NaN(13514, 62.2 %) 1.0-998.0
14 n_rooms                      i 0-24 uniq(19) TOP3: 3(7214); 2(5499); 4(3367);  BOTTOM1 16(1); 
15 n_bathrooms                  f NaN(16, 0.1 %) 1.0-16.

### Bool type features

In [ ]:
i = 1

notes_inline = {
    '' : '',
}
notes_below = {
    '' : '',
}
#TODO: mixed tipus...
for split in [True, False]:
  for col in df.columns:
    feature = col
    if df[col].dtype in ['bool'] and 'id' != col:
      nan = df[feature].isna().value_counts()
      if split == (True in nan and nan[True] == len(df)):
        print(str(i).ljust(2), col.ljust(28),  str(df[col].dtype)[0], end='')
        BOLD, END = '\033[1m', '\033[0m'
        #display(df[feature].head())
        
        if True in nan:
          print(' '+BOLD + 'NaN('+str(nan[True])+', '+percent_from_total(nan[True])+')' + END, end='')
        
        print(' ', end='')
        for val, cnt in df[feature].value_counts().items():
          print(val, '(', cnt, ', ', percent_from_total(cnt), '); ', sep='', end='')
        print()
        i += 1

1  is_exact_address_hidden      b True(15002, 69.0 %); False(6740, 31.0 %); 
2  is_rent_price_known          b False(21742, 100.0 %); 
3  is_buy_price_known           b True(21742, 100.0 %); 
4  is_renewal_needed            b False(17747, 81.6 %); True(3995, 18.4 %); 
5  has_parking                  b False(14023, 64.5 %); True(7719, 35.5 %); 


### Mixed and string type features

In [ ]:
len(df.columns)

In [ ]:
i = 1

notes_inline = {
    '' : '',
}
notes_below = {
    '' : '',
}
#TODO: mixed tipus...

for col in df.columns:
  feature = col
  if df[col].dtype in ['object'] and 'id' != col:
    if (col.startswith('is_') or col.startswith('has_')):
      nan = df[feature].isna().value_counts()
      print(i, ' ', col, end='')
      #if split == (True in nan and nan[True] == len(df)):
      #print(str(i).ljust(2), col.ljust(28),  str(df[col].dtype)[0], end='')
      BOLD, END = '\033[1m', '\033[0m'
      #display(df[feature].head())
      
      if True in nan:
        print(' '+BOLD + 'NaN('+str(nan[True])+', '+percent_from_total(nan[True])+')' + END, end='')
        
      print(' ', end='')
      #for val, cnt in df[feature].value_counts().items():
      #  print(val, '(', cnt, ', ', percent_from_total(cnt), '); ', sep='', end='')
      print('uniq('+BOLD+str(len(df[col].value_counts()))+END+ ')', 
  ' TOP 3', str(df[col].value_counts().nlargest(3)).replace('\n', '; ')
  #, 'BOTTOM 1', df[col].value_counts().nsmallest(1)
  )
      print()
      i += 1

1   is_floor_under NaN(1170, 5.4 %) uniq(2)  TOP 3 False    18033; True      2539; Name: is_floor_under, dtype: int64

2   is_new_development NaN(992, 4.6 %) uniq(2)  TOP 3 False    19281; True      1469; Name: is_new_development, dtype: int64

3   has_central_heating NaN(8134, 37.4 %) uniq(2)  TOP 3 False    9516; True     4092; Name: has_central_heating, dtype: int64

4   has_individual_heating NaN(8134, 37.4 %) uniq(2)  TOP 3 True     9516; False    4092; Name: has_individual_heating, dtype: int64

5   has_ac NaN(10531, 48.4 %) uniq(1)  TOP 3 True    11211; Name: has_ac, dtype: int64

6   has_fitted_wardrobes NaN(8343, 38.4 %) uniq(1)  TOP 3 True    13399; Name: has_fitted_wardrobes, dtype: int64

7   has_lift NaN(2386, 11.0 %) uniq(2)  TOP 3 True     14895; False     4461; Name: has_lift, dtype: int64

8   is_exterior NaN(3043, 14.0 %) uniq(2)  TOP 3 True     16922; False     1777; Name: is_exterior, dtype: int64

9   has_garden NaN(20186, 92.8 %) uniq(1)  TOP 3 True    1556; Name:

In [ ]:
len(df[df.rent_price < 1].rent_price)
df[df.rent_price < 1].rent_price.value_counts()

-2333       80
-51191      75
-1225       74
-104074     71
-38920      67
            ..
-109753      1
-3353444     1
-110916      1
-18887       1
-1027        1
Name: rent_price, Length: 299, dtype: int64

In [ ]:
['# of uniq ', len(df['n_floors'].value_counts()), 
 ' TOP 3', str(df['n_floors'].value_counts().nlargest(3)).replace('\n', '; '), 
 'BOTTOM 1', df['n_floors'].value_counts().nsmallest(1)]
#pandas.core.series.Series
#type(df['n_floors'].value_counts().nlargest(3))

#for val, cnt in df['n_floors'].value_counts().nlargest(3).items():
#  print(val, '(', cnt, '); ', sep='', end='')

['# of uniq ',
 6,
 ' TOP 3',
 '4.00    567; 3.00    476; 2.00    316; Name: n_floors, dtype: int64',
 'BOTTOM 1',
 7.00    1
 Name: n_floors, dtype: int64]

## Visas feature'ų aprašymas

In [ ]:
'''
#21742 rows 
| Variable Name  | Description            | Example
| -----------    | -----------            | 
| ID             | Identifier ID          |
| title         o| Title from listing     | Piso en venta en calle de Godella, 64 -> Butas parduodamas adresu Calle de Godella, 64
| subtitle      o| Neighborhood and city  | San Cristóbal, Madrid
| sq_mt_built   f| Square meter built     | 65- 770m2, NaN(126, XX%), reikšmių(21616, XX%)
| sq_mt_useful  f| Square meter useful    | 16 - 778m2, NaN(13514), kai su reikšmėm - 8228 
| n_rooms       i| Number of rooms        | 1-24
| n_bathrooms   f| Number of bathrooms    | 1-16
| n_floors      f|                        | NaN(20305), 1-7. Išskyrus reikšmę 6.
| sq_mt_allotment f| Square meter allotment | 1-382, NaN(20310), kai su reikšmėm - 1432

- | latitude      f| Latitude (Since portals hide the location mainly NAs)  | 21742- NaN, NĖRA reikšmių
- | longitude     f| Longitude (Since portals hide the location mainly NAs) | 21742- NaN, NĖRA reikšmių
| raw_address   o| Address                | NaN(5465), kai su rekšmėm (16277)
| is_exact_address_hidden b| Boolean values| True-15002; False-6740 -> NaN(15002), kai su rekšmėm (6740)
| street_name   o| street name            | NaN(5905)  , kai su reikšmėm - 15837
| street_number o| street number          | 23 - A, NaN(15442), kai su reikšmėm - 6300
-| portal       f|                        | NaN(21742), NĖRA reikšmių
| floor         o|                        | NaN(2607), kai su reikšmėm  - 19135
  1                       4440
  2                       3546
  3                       3001
  4                       2323
  Bajo                    2144 #Apatinis aukštas arba namo aukštas, esantis gatvės lygyje.
  5                       1310
  6                        913
  7                        556
  8                        326
  Entreplanta exterior     236 #Floor plan of a house or building that is built by removing part of the height from the lower or upper floor.
  9                        181
  Semi-sótano exterior      55 | pusrūsio eksterjeras
  #sótano=rūsys
  <...>
| is_floor_under o|                        | False-18033; True-2539 -> NAN(1170), kai su reikšmėm - 20572
-| door          f|                        | NaN(21742), NĖRA reikšmių


 buy_price             | Target Value    |

Kęstutis:
| neighborhood_id     o| | #126 Neighborhood 23: Malasaña-Universidad (5196.25 €/m2) - District 4: Centro                485           
  Tas pats mikrorajonas/District 2
  Neighborhood 9: Campo de las Naciones-Corralejos (3417.44 €/m2) - District 2: Barajas      9
  Neighborhood 10: Casco Histórico de Barajas (3100.87 €/m2) - District 2: Barajas           6
  Neighborhood 11: Timón (2879.88 €/m2) - District 2: Barajas                                6
| operation           o| | sale - #21742
| rent_price          i| | int visi yra: -34590276-18421 #TODO: clean. Dig deeper
-| rent_price_by_area f| | NĖRA
-| is_rent_price_known b| | NĖRA
| buy_price           i| | 36000 - 8800000
| buy_price_by_area   i| | 447-18889. #5643 of uniq. Paplitimas: 5000    143, 3000     99, ...
| is_buy_price_known  b| | visi True NEREIKALINGA
| house_type_id       o|  | NaN(391)
  HouseType 1: Pisos            17705
  HouseType 2: Casa o chalet     1938 # namas ar namelis| namas ar kotedžas
  HouseType 5: Áticos            1032 # mansardiniai namai. Gal namai su mansardom?
  HouseType 4: Dúplex             676 # Dvibutis namas, bendra siena.
| is_renewal_needed  b| | False    17747, True      3995
| is_new_development o| | NaN(992). False    19281, True      1469
| built_year         f| | NaN(11742)
| has_central_heating o| | NaN(8134). False    9516, True     4092
-| has_individual_heating o| NaN(8134): True     9516, False    4092 //nes vienas kitą paneigia has_central_heating
-| are_pets_allowed   f| | **NĖRA** reikšmių
| has_ac             o| | NaN(10531), True    11211, False reikšmių nėra!!!
| has_fitted_wardrobes o| | NaN(8343), True    13399, False reikšmių nėra!!!
| has_lift           o| | NaN(2386), True     14895, False     4461
| is_exterior        o| | NaN(3043), True     16922, False     1777

Alina:
| has_garden      o| Garden                 | NaN (20186), kai su reikšmėm 1556, True - 1556 |
| has_pool        o| Pool                   | NaN (16571), kai su reikšmėm 5171, True - 5171 |
| has_terrace     o| Terrace                | NaN (12194), kai su reikšmėm 9548, True - 9548 |
| has_balcony     o| Balcony                | NaN (18421), kai su reikšmėm 3321, True - 3321 |
| has_storage_room o| Storage room           | NaN (14044), kai su reikšmėm 7698, True - 7698 |
-| is_furnished    f| Is furnished           | NaN (21742)                       |
-| is_kitchen_equipped f| Kitchen is equipped | NaN (21742)                       |
| is_accessible   o| Accessible             | NaN (17668), kai su reikšmėm 4074, True - 4074 |
| has_green_zones o| Green zones            | NaN (17685), kai su reikšmėm 4057, True - 4057 |
| energy_certificate o| Energy certificate  | en trámite(10937), no indicado (3669), E(2701), D(1121), G(894), F(674), A(618), C(580), B(452), inmueble exento(96)|
| has_parking     b| Parking                | True - 7719, False - 14023                     |
-| has_private_parking f| Private parking    | NaN (21742)                       |
-| has_public_parking  f| Public parking     | NaN (21742)                       |
| is_parking_included_in_price o| Parking is included in price | NaN (14023), kai su reikšmėm 7719, True - 7115, False - 604 |
| parking_price   f| Parking price          | 0-30000, NaN (14023), kai su reikšmėm 7719|
| is_orientation_north o| North orientation | NaN (10384), kai su reikšmėm 10384, True - 2791, False - 8567|
| is_orientation_west  o| West orientation  | NaN (10384), kai su reikšmėm 10384, True - 4071, False - 7287|
| is_orientation_south o| South orientation | NaN (10384), kai su reikšmėm 10384, True - 5823, False - 5535|
| is_orientation_east  o| East orientation  | NaN (10384), kai su reikšmėm 10384, True - 5056, False - 6302|
'''
print()

## Boxplots of the 6 selected features

In [ ]:
def boxplot(data, size=(6.4, 4.8)):
  sns.set(rc={"figure.figsize":size})
  sns.boxplot(data=data)
  plt.grid()
  plt.show()

feature = 'buy_price'
print(feature)
boxplot(df[feature])
'''
feature = 'n_floors'
print(feature)
boxplot(df['n_floors'])
'''
features = ['sq_mt_built', 'sq_mt_useful', 'n_rooms', 'n_bathrooms', 'buy_price_by_area', 'has_parking']
for feature in features:
  print(feature)
  boxplot(df[feature])

buy_price
sq_mt_built
sq_mt_useful
n_rooms
n_bathrooms
buy_price_by_area
has_parking


# MAE, MAPE, WMAPE, RMSE, MSE 

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.metrics import mean_absolute_percentage_error, mean_squared_error, mean_absolute_error

def mae(y_test, y_pred):
    """mean absolute error"""
    return round(mean_absolute_error(y_test, y_pred), 3)

def rmse(y_test, y_pred):
    """root mean squared error"""
    return round(np.sqrt(mean_squared_error(y_test, y_pred)), 4)

def mape(y_test, y_pred):
    """mean absolute percentage error"""
    return round(mean_absolute_percentage_error(y_test, y_pred), 4)

def wmape(y_test, y_pred):
    """weighted mean absolute percentage error"""
    wmape = sum(np.abs(y_test-y_pred)) / sum(np.abs(y_test))
    return round(wmape, 4)

def mse(y_test, y_pred):
    """mean squared error"""
    return round(mean_squared_error(y_test, y_pred), 4)

# TODO: Kokios paklaidos/tikslumas skirtinguose rėžiuose

## a_results_tbls

In [ ]:
#Model	Details	RMSE (train)	RMSE (test)
columns = ['', 'MAE tr', 'MAE te',
            'MAPE tr', 'MAPE te',
            'wMAPE tr', 'wMAPE te',
            'RMSE tr', 'RMSE te', 
           #'MSE tr', 'MSE te', 
           'len', 'R^2']
new_DF = pd.DataFrame(columns=columns)#MSE?
#df[['MAE tr', 'MAPE tr', 'wMAPE tr', 'RMSE tr']] = 'background-color: grey'

#print(type(new_DF))
a_result_tbls = []
for i in range(10):
  a_result_tbls.append({
    'model': '',
    'details': str(i*100)+'k-'+str(i*100+100)+'k',# _000 _999
    'tbl': new_DF.copy()
  })

for i in range(10):
  item = a_result_tbls[i]
  print(BOLD + item['model'] + END + item['details'])
  display(item['tbl'])

0k-100k


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

100k-200k


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

200k-300k


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

300k-400k


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

400k-500k


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

500k-600k


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

600k-700k


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

700k-800k


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

800k-900k


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

900k-1000k


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

## Data ranges (step = 100k €)

In [ ]:
df_WO_NaN = df[(~df['sq_mt_built'].isna()) & (~df['sq_mt_useful'].isna()) & (~df['n_rooms'].isna()) & (~df['n_bathrooms'].isna()) & (~df['buy_price_by_area'].isna()) & (~df['has_parking'].isna()) & (~df['buy_price'].isna())]

a_model_dfs = []
for i in range(10):
  nuo = i*100_000
  iki = i*100_000+100_000
  idx_no_out =  (nuo <= df_WO_NaN.buy_price) & (df_WO_NaN.buy_price < iki)
  a_model_dfs.append(df_WO_NaN[idx_no_out].copy())

a_model_dfs[0].buy_price
#NUO <= x < IKI
#(NUO <= x) & (x < IKI)

0        85000
36       96000
39       92500
46       85000
106      89900
         ...  
20853    69000
21148    90000
21213    99000
21217    99950
21345    99900
Name: buy_price, Length: 206, dtype: int64

## data train test split and model train

In [ ]:
features = ['sq_mt_built', 'sq_mt_useful', 'n_rooms', 'n_bathrooms', 'buy_price_by_area', 'has_parking']
for i in range(10):
  df_selected = a_model_dfs[i]
  x = df_selected[features]
  y = df_selected.buy_price

  # Let's use 70% of random sample to train the model, and the remaining 30% of sample to test how good is the model
  X_train,X_test,y_train,y_test=train_test_split(x, y, train_size = 0.7, random_state=42)

  lr = LinearRegression()
  #display(X_train.head())
  reg = lr.fit(X_train, y_train)
  #lr???
  y_pred_train = reg.predict(X_train)
  y_pred_test = reg.predict(X_test)

  #round(np.sqrt(mean_squared_error(y_train, y_pred_train)),3)
  #round(np.sqrt(mean_squared_error(y_test, y_pred)),3)
  round_to = 2
  mae_tr = round(mae(y_train, y_pred_train), round_to)
  mae_te = round(mae(y_test, y_pred_test), round_to)

  mape_tr = round(mape(y_train, y_pred_train), round_to)
  mape_te = round(mape(y_test, y_pred_test), round_to)

  wmape_tr = round(wmape(y_train, y_pred_train), round_to)
  wmape_te = round(wmape(y_test, y_pred_test), round_to)

  rmse_tr = round(rmse(y_train, y_pred_train), round_to)
  rmse_te = round(rmse(y_test, y_pred_test), round_to)

  #mse_tr = round(mse(y_train, y_pred_train), round_to)
  #mse_te = round(mse(y_test, y_pred_test), round_to)
  '''
  print(f'MAE: {mae(y_test, y_pred)}')
  print(f'MAPE: {mape(y_test, y_pred)}')
  print(f'WMAPE: {wmape(y_test, y_pred)}')
  print(f'RMSE: {rmse(y_test, y_pred)}')
  print(f'MSE: {mse(y_test, y_pred)}')
'MAE tr', 'MAE te', 'MAPE tr', 'MAPE te','wMAPE tr', 'wMAPE te', 'RMSE tr', 'RMSE te', 'MSE tr', 'MSE te'
  '''

  import statsmodels.api as sm
  pd.options.mode.chained_assignment = 'warn'#None
  x['has_parking'] = x['has_parking'].astype(int)#.copy(deep=True)
  #x.loc[:, 'has_parking'] = x['has_parking'].astype(int).copy(deep=True)
  #display(x['has_parking'].value_counts())
  x_train = sm.add_constant(x, prepend=False)
  #print(np.asarray(y))
  #display(x_train)
  #print(np.asarray(x_train))
  mod = sm.OLS(y, x_train)#y, x_train
  res = mod.fit()
  lines = res.summary()
  #print(type(lines))
  #<class 'statsmodels.iolib.summary.Summary'>
  lines = str(lines).split('\n')
  Rsquared = float(lines[2].split('R-squared:')[1].strip())
  #print(Rsquared)

  a_result_tbls[i]['tbl'].loc[0] = [a_result_tbls[i]['details'], mae_tr, mae_te, mape_tr, mape_te, wmape_tr, wmape_te, rmse_tr, rmse_te, len(y), Rsquared]
  #mse_tr, mse_te, 

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:45: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:117: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:45: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:117

https://www.statsmodels.org/dev/regression.html#examples
https://www.statsmodels.org/dev/generated/statsmodels.tools.tools.add_constant.html

---

endog: array_like    
>  A 1-d endogenous response variable. The dependent variable.

exog: array_like    
>  A **nobs** x **k** array where **nobs** is the number of observations and **k** is the number of regressors. An intercept is not included by default and should be added by the user. See statsmodels.tools.add_constant.

https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html

##Show/Display/print prediction errors table (MAE, MAPE, wMAPE, RMSE, R^2)

In [ ]:
print("LinearRegression")
def highlight_cols(s):
    color = 'yellow'
    return 'background-color: %s' % color

#pd.set_option('display.float_format', lambda x: '%9.2f' % x)
pd.options.display.float_format = '{:20.2f}'.format
for i in range(10):
  item = a_result_tbls[i]
  if item['model']:
    print(BOLD + item['model'] + END)# + item['details'])
  i_DF = item['tbl'].copy()
  #i_DF[['MAE te', 'MAPE te', 'wMAPE te', 'RMSE te']] = 'background-color: yellow'
  #i_DF = df.style.apply(i_DF, axis = None)
  
  i_DF = i_DF.style.applymap(highlight_cols, subset=pd.IndexSlice[:, ['MAE te', 'MAPE te', 'wMAPE te', 'RMSE te']])
  i_DF = i_DF.format({  'MAE tr':'{:9.2f}', 'MAE te':'{:9.2f}',
                               'MAPE tr':'{:9.2f}', 'MAPE te':'{:9.2f}',
                               'wMAPE tr':'{:9.2f}', 'wMAPE te':'{:9.2f}',
                               'RMSE tr':'{:9.2f}', 'RMSE te':'{:9.2f}'})
                              #  'MSE tr':'{:9.2f}', 'MSE te':'{:9.2f}'})
  #i_DF = i_DF.format_index(precision=3) 
  #print(type(i_DF))

  #columns
  #display(i_DF)
  import IPython
  #/*.col_heading row0 */
  style = '''<style>
.data {
  width: 68px;
  #background-color: yellow;
}
td {text-align: right}
tr:hover {background-color: yellow;}
</style>'''
  style_hide_header = ''
  #if i != 0:
  #  style_hide_header = '''<style>#thead {display: none}</style>'''
#/*68px;*/
#  style_hide_header = '''<style>:not(thead:first-of-type) {display: none}</style>'''#thead {display: none}

  html = i_DF.to_html()
  if i != 0:
    html = html.replace('thead', 'thead hidden')
  display(IPython.display.HTML(html+style+style_hide_header))
  #display(IPython.display.HTML(i_DF.to_html()+style))
  #print(i_DF.to_html())

LinearRegression


# Data plots. How sq_mt influence price_per_sq_mt in different ranges of 100k €

In [ ]:
def plot_fit_var(X_train,X_test,y_train,y_test, y_pred_train, y_pred_test, var):
  fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(10, 5))

  print(type(X_train[var].min()))
  print(X_test[var].min())
  X_min = min(X_train[var].min(), X_test[var].min())
  X_max = max(X_train[var].max(), X_test[var].max())

  y_min = min(min(y_train), min(y_test), min(y_pred_train), min(y_pred_test))
  y_max = max(max(y_train), max(y_test), max(y_pred_train), max(y_pred_test))

  #print(type(axes[0]))
  axes[0].set_xlim([X_min, X_max])
  axes[0].set_ylim([y_min, y_max])
  axes[1].set_xlim([X_min, X_max])
  axes[1].set_ylim([y_min, y_max])

  axes[0].scatter(X_train[var], y_train, color='gray', s=10, alpha=0.3)
  axes[0].scatter(X_train[var], y_pred_train, s=10, color='red', alpha=0.3)
  axes[0].plot([[X_train[var].median()], [X_train[var].median()]], [y_min, y_max], color='green', alpha=0.3)
  axes[0].plot([[X_min], [X_max]], [y_train.median(), y_train.median()], color='green', alpha=0.3)
  #X_train.median()
  #axes[0].set_xlim(bottom=X_min, top=X_max)
      #set_xlim([0, 5])
  #axes[0].set_ylim(bottom=y_min, top=y_max)
  
  
  axes[1].scatter(X_test[var], y_test, color='gray', s=10, alpha=0.3)
  axes[1].scatter(X_test[var], y_pred_test, color='red', s=10, alpha=0.3)
  axes[1].plot([[X_test[var].median()], [X_test[var].median()]], [y_min, y_max], color='green', alpha=0.3)
  axes[1].plot([[X_min], [X_max]], [y_test.median(), y_test.median()], color='green', alpha=0.3)

  axes[0].title.set_text('Train Sample')
  axes[1].title.set_text('Test Sample')
  fig.tight_layout()
  plt.show()

In [ ]:
# features = ['sq_mt_built', 'sq_mt_useful', 'n_rooms', 'n_bathrooms', 'buy_price_by_area', 'has_parking']
features2 = ['sq_mt_built']# 'sq_mt_built', 'sq_mt_useful', 'n_rooms', 'n_bathrooms', 'buy_price_by_area', 'has_parking']
for i in range(10):
  df_selected = a_model_dfs[i]
  x = df_selected[features2]
  y = df_selected.buy_price_per_sq_mt_built

  # Let's use 70% of random sample to train the model, and the remaining 30% of sample to test how good is the model
  X_train,X_test,y_train,y_test=train_test_split(x, y, train_size = 0.7, random_state=42)

  lr = LinearRegression()
  #display(X_train.head())
  reg = lr.fit(X_train, y_train)
  
  y_pred_train = reg.predict(X_train)
  y_pred_test = reg.predict(X_test)
  x_axis_var = 'sq_mt_built'#'sq_mt_built'
  display(X_train['sq_mt_built'].min())
  plot_fit_var(X_train,X_test,y_train,y_test, y_pred_train, y_pred_test, x_axis_var)

18.0

<class 'float'>
13.0


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  


TypeError: ignored

In [ ]:
#min(5, 1)
type(min)
min


13.0

# How `sq_mt_built` influnce `buy_price_per_sq_mt_built`

In [ ]:
#def plot_fit_var(X_train,X_test,y_train,y_test, y_pred_train, y_pred_test, var):
x_axis_var = 'sq_mt_built'
y_axis_var = 'buy_price_per_sq_mt_built'
df = full_df
df.head()

upper_price_limit =  500_000

df = df[df.buy_price < upper_price_limit]
display(df[x_axis_var].isna().value_counts())
df = df[(~df['sq_mt_built'].isna())]
display(df[x_axis_var].isna().value_counts())


lr = LinearRegression()
reg = lr.fit(df[x_axis_var].values.reshape(-1, 1), df[y_axis_var].values)
y_pred = reg.predict(df[x_axis_var].values.reshape(-1, 1))

fig, axes = plt.subplots(nrows=1, ncols=1)#, figsize=(10, 5)
X_min = min(df[x_axis_var])
X_max = max(df[x_axis_var])
y_min = min(df[y_axis_var])
y_max = max(df[y_axis_var])
axes.set_xlim([X_min, X_max])
axes.set_ylim([y_min, y_max])

x_limit = []
y_limit = []
for x in range(int(X_min), int(X_max)):
  x_limit.append(x)
  y_limit.append(upper_price_limit/x)

axes.scatter(df[x_axis_var], df[y_axis_var], color='gray', s=1, alpha=0.3)
axes.scatter(df[x_axis_var], y_pred, color='red', s=10, alpha=0.3)
axes.plot(x_limit, y_limit, color='blue', alpha=0.3)
#axes[0].set_xlim(bottom=X_min, top=X_max)
    #set_xlim([0, 5])
#axes[0].set_ylim(bottom=y_min, top=y_max)
#axes[0].title.set_text('Train Sample')
fig.tight_layout()
plt.show()

In [ ]:
df.buy_price_per_sq_mt_built.median()

# Histograms

In [ ]:
def plot_hist(x, bins='auto', binrange=(0, 1_000_000)):
    plt.figure(figsize=(12,5))
    plt.tight_layout()
    sns.histplot(x, bins=bins, binrange=binrange)
    plt.show()


## Histogram: full data set. How many apartments in price ranges.

In [ ]:
print(len(full_df))
plot_hist(full_df['buy_price'], 40, binrange=(0, 8_000_000))

21742


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  


## Histogram: W/O NaN values of 6 selected features, data limited by buy_price to 1_000_000. How many apartments are in 100k price ranges.

In [ ]:
print(len(df_WO_NaN))
# df_WO_NaN.head()

8206


In [ ]:
y = df_WO_NaN.buy_price 
idx_no_out = df_WO_NaN.buy_price < 1_000_000
features = ['sq_mt_built', 'sq_mt_useful', 'n_rooms', 'n_bathrooms', 'buy_price_by_area', 'has_parking']
x = df_WO_NaN[features] 
y = y[idx_no_out].copy()
x = x[idx_no_out].copy()

In [ ]:
print(len(y))
plot_hist(y, 10)

7519


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  


## Histogram. sq_mt_built: range of 10.



In [ ]:
# sq_mt_built
min, max = full_df['sq_mt_built'].min(),full_df['sq_mt_built'].max()
plot_hist(full_df['sq_mt_built'], 39, binrange=(min,400))
(full_df['sq_mt_built'].min(),full_df['sq_mt_built'].max())
#10

#Bendrabučio 16m^2
#33-37 buto 1 kambario
#44-46 2 kambarių
#64 - 3 kamb.
# Maks. 60 m^2.

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  


(13.0, 999.0)

###a_results_tbls

In [ ]:
columns = ['', 'MAE tr', 'MAE te',
                               'MAPE tr', 'MAPE te',
                               'wMAPE tr', 'wMAPE te',
                               'RMSE tr', 'RMSE te', 'len', 'R^2']
new_DF = pd.DataFrame(columns=columns)#MSE?
#df[['MAE tr', 'MAPE tr', 'wMAPE tr', 'RMSE tr']] = 'background-color: grey'

#print(type(new_DF))
a_result_tbls = []
for i in range(1, 40):
  a_result_tbls.append({
    'model': '',
    'details': str(i*10)+'-'+str(i*10+10)+'m^2',# _000 _999
    'tbl': new_DF.copy()
  })

for i in range(10):
  item = a_result_tbls[i]
  print(BOLD + item['model'] + END + item['details'])
  display(item['tbl'])

10-20m^2


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

20-30m^2


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

30-40m^2


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

40-50m^2


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

50-60m^2


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

60-70m^2


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

70-80m^2


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

80-90m^2


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

90-100m^2


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

100-110m^2


Empty DataFrame
Columns: [, MAE tr, MAE te, MAPE tr, MAPE te, wMAPE tr, wMAPE te, RMSE tr, RMSE te, len, R^2]
Index: []

### Data ranges (step = 10m^2)

In [ ]:
df_WO_NaN = df[(~df['sq_mt_built'].isna()) & (~df['sq_mt_useful'].isna()) & (~df['n_rooms'].isna()) & (~df['n_bathrooms'].isna()) & (~df['buy_price_by_area'].isna()) & (~df['has_parking'].isna()) & (~df['buy_price'].isna())]

a_model_dfs = []
for i in range(1, 40):
  nuo = i*10
  iki = i*10+10
  idx_no_out =  (nuo <= df_WO_NaN.sq_mt_built) & (df_WO_NaN.sq_mt_built < iki)
  a_model_dfs.append(df_WO_NaN[idx_no_out].copy())

a_model_dfs[0].sq_mt_built
#NUO <= x < IKI
#(NUO <= x) & (x < IKI)

3248                   18.00
5650                   13.00
18997                  18.00
Name: sq_mt_built, dtype: float64

### data train test split and model train

In [ ]:
features = ['sq_mt_built', 'sq_mt_useful', 'n_rooms', 'n_bathrooms', 
            #'buy_price_by_area', 
            'has_parking']
for i in range(20):
  df_selected = a_model_dfs[i]
  x = df_selected[features]
  y = df_selected.buy_price

  # Let's use 70% of random sample to train the model, and the remaining 30% of sample to test how good is the model
  X_train,X_test,y_train,y_test=train_test_split(x, y, train_size = 0.7, random_state=42)

  lr = LinearRegression()
  #display(X_train.head())
  reg = lr.fit(X_train, y_train)
  #lr???
  y_pred_train = reg.predict(X_train)
  y_pred_test = reg.predict(X_test)

  #round(np.sqrt(mean_squared_error(y_train, y_pred_train)),3)
  #round(np.sqrt(mean_squared_error(y_test, y_pred)),3)
  round_to = 2
  mae_tr = round(mae(y_train, y_pred_train), round_to)
  mae_te = round(mae(y_test, y_pred_test), round_to)

  mape_tr = round(mape(y_train, y_pred_train), round_to)
  mape_te = round(mape(y_test, y_pred_test), round_to)

  wmape_tr = round(wmape(y_train, y_pred_train), round_to)
  wmape_te = round(wmape(y_test, y_pred_test), round_to)

  rmse_tr = round(rmse(y_train, y_pred_train), round_to)
  rmse_te = round(rmse(y_test, y_pred_test), round_to)
  '''
  print(f'MAE: {mae(y_test, y_pred)}')
  print(f'MAPE: {mape(y_test, y_pred)}')
  print(f'WMAPE: {wmape(y_test, y_pred)}')
  print(f'RMSE: {rmse(y_test, y_pred)}')
'MAE tr', 'MAE te', 'MAPE tr', 'MAPE te','wMAPE tr', 'wMAPE te', 'RMSE tr', 'RMSE te'
  '''

  import statsmodels.api as sm
  pd.options.mode.chained_assignment = 'warn'#None
  x['has_parking'] = x['has_parking'].astype(int)#.copy(deep=True)
  #x.loc[:, 'has_parking'] = x['has_parking'].astype(int).copy(deep=True)
  #display(x['has_parking'].value_counts())
  x_train = sm.add_constant(x, prepend=False)
  #print(np.asarray(y))
  #display(x_train)
  #print(np.asarray(x_train))
  mod = sm.OLS(y, x_train)#y, x_train
  res = mod.fit()
  lines = res.summary()
  #print(type(lines))
  #<class 'statsmodels.iolib.summary.Summary'>
  lines = str(lines).split('\n')
  Rsquared = round(float(lines[2].split('R-squared:')[1].strip()), 3)
  #print(Rsquared)

  a_result_tbls[i]['tbl'].loc[0] = [a_result_tbls[i]['details'], mae_tr, mae_te, mape_tr, mape_te, wmape_tr, wmape_te, rmse_tr, rmse_te, len(y), Rsquared]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:43: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:117: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)
/usr/local/lib/python3.7/dist-packages/statsmodels/stats/stattools.py:71: ValueWarning: omni_normtest is not valid with less than 8 observations; 3 samples were given.
  "samples were given." % int(n), ValueWarning)
/usr/local/lib/python3.7/dist-packages/statsmodels/regression/linear_model.py:1648: RuntimeWarning: divide by zero encountered in true_divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resi

###Show/Display/print prediction errors table (MAE, MAPE, wMAPE, RMSE, R^2)

In [ ]:
print("LinearRegression")
def highlight_cols(s):
    color = 'yellow'
    return 'background-color: %s' % color

#pd.set_option('display.float_format', lambda x: '%9.2f' % x)
pd.options.display.float_format = '{:20.2f}'.format
for i in range(39):
  item = a_result_tbls[i]
  if item['model']:
    print(BOLD + item['model'] + END)# + item['details'])
  i_DF = item['tbl'].copy()
  #i_DF[['MAE te', 'MAPE te', 'wMAPE te', 'RMSE te']] = 'background-color: yellow'
  #i_DF = df.style.apply(i_DF, axis = None)
  
  i_DF = i_DF.style.applymap(highlight_cols, subset=pd.IndexSlice[:, ['MAE te', 'MAPE te', 'wMAPE te', 'RMSE te']])
  i_DF = i_DF.format({  'MAE tr':'{:9.2f}', 'MAE te':'{:9.2f}',
                               'MAPE tr':'{:9.2f}', 'MAPE te':'{:9.2f}',
                               'wMAPE tr':'{:9.2f}', 'wMAPE te':'{:9.2f}',
                               'RMSE tr':'{:9.2f}', 'RMSE te':'{:9.2f}',
                               'R^2': '{:9.2f}'})
  #i_DF = i_DF.format_index(precision=3) 
  #print(type(i_DF))

  #columns
  #display(i_DF)
  import IPython
  #/*.col_heading row0 */
  style = '''<style>
.data {
  width: 68px;
  #background-color: yellow;
}
td {text-align: right}
tr:hover {background-color: yellow;}
</style>'''
  style_hide_header = ''
  #if i != 0:
  #  style_hide_header = '''<style>#thead {display: none}</style>'''
#/*68px;*/
#  style_hide_header = '''<style>:not(thead:first-of-type) {display: none}</style>'''#thead {display: none}

  html = i_DF.to_html()
  if i != 0:
    html = html.replace('thead', 'thead hidden')
  display(IPython.display.HTML(html+style+style_hide_header))
  #display(IPython.display.HTML(i_DF.to_html()+style))
  #print(i_DF.to_html())

LinearRegression


#Scatter plots

## buy_price vs. 6 selected int features

In [ ]:
sns.pairplot(data=full_df, x_vars=['sq_mt_built', 'sq_mt_useful', 'n_rooms'], y_vars=["buy_price"], height=5, plot_kws={'s':15, 'alpha': 0.3})
sns.pairplot(data=full_df, x_vars=['n_bathrooms', 'buy_price_by_area', 'has_parking'], y_vars=["buy_price"], height=5, plot_kws={'s':15, 'alpha': 0.3})

/usr/local/lib/python3.7/dist-packages/seaborn/axisgrid.py:1242: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  fig = plt.figure(figsize=figsize)
/usr/local/lib/python3.7/dist-packages/seaborn/axisgrid.py:1242: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  fig = plt.figure(figsize=figsize)


## buy_price (limited) vs. selected features

In [ ]:
sns.pairplot(data=df_WO_NaN[df_WO_NaN.buy_price < 1_000_000], x_vars=['sq_mt_built', 'sq_mt_useful', 'n_rooms'], y_vars=['buy_price'], height=5, plot_kws={'s':15, 'alpha': 0.3})
sns.pairplot(data=df_WO_NaN[df_WO_NaN.buy_price < 1_000_000], x_vars=['n_bathrooms', 'buy_price_by_area', 'has_parking'], y_vars=['buy_price'], height=5, plot_kws={'s':15, 'alpha': 0.3})

/usr/local/lib/python3.7/dist-packages/seaborn/axisgrid.py:1242: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  fig = plt.figure(figsize=figsize)
/usr/local/lib/python3.7/dist-packages/seaborn/axisgrid.py:1242: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  fig = plt.figure(figsize=figsize)


# Correlation matrix: selected features 

## full data set

In [ ]:
features = ['sq_mt_built', 'sq_mt_useful', 'n_rooms', 'n_bathrooms', 'buy_price_by_area', 'has_parking', 'buy_price']
corr_matrix = full_df[features].corr().round(3)
corr_plot(corr_matrix, "Correlation matrix")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  


## data set limited by buy_price

In [ ]:
limited_df = df_WO_NaN[df_WO_NaN.buy_price < 1_000_000]
corr_matrix = limited_df[features].corr().round(3)
corr_plot(corr_matrix, "Correlation matrix")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  


# Linear Regression: MAE, MAPE, WMAPE, RMSE. W/O NaN values of 6 selected features, data limited by buy_price to 1_000_000

In [ ]:
len(x), len(y), len(limited_df)

(100, 100, 7519)

In [ ]:
# Let's use 70% of random sample to train the model, and the remaining 30% of sample to test how good is the model
X_train,X_test,y_train,y_test=train_test_split(x, y, train_size = 0.7, random_state=42)

lr = LinearRegression()
reg = lr.fit(X_train, y_train)
y_pred = reg.predict(X_test)

print(f'MAPE: {mape(y_test, y_pred)}')
print(f'WMAPE: {wmape(y_test, y_pred)}')
print(f'MAE: {mae(y_test, y_pred)}')
print(f'RMSE: {rmse(y_test, y_pred)}')

MAPE: 0.2284
WMAPE: 0.2232
MAE: 193767.911
RMSE: 264910.6049


# log-linear model: MAE, MAPE, WMAPE, RMSE. W/O NaN values of 6 selected features, data limited by buy_price to 1_000_000


In [ ]:
y_train_log = np.log(y_train)
lr_log = LinearRegression()
reg_log = lr_log.fit(X_train, y_train_log)
y_pred_log = reg_log.predict(X_test)
y_pred_v2 = np.exp(y_pred_log)

print(f'MAPE: {mape(y_test, y_pred_v2)}')
print(f'WMAPE: {wmape(y_test, y_pred_v2)}')
print(f'MAE: {mae(y_test, y_pred_v2)}')
print(f'RMSE: {rmse(y_test, y_pred_v2)}')

MAPE: 0.2319
WMAPE: 0.2263
MAE: 196421.117
RMSE: 274535.045


In [ ]:
'''
- Reikia iš turimų žinių surašyti kintamuosius, kurie įtakoja kainą.
- Reikia iš duomenų duotus kintamuosius patikrinti:
  - NaN
  - ar netušti
  - ar jie koreliuoja su kaina
  - ar nėra outlier'ių
  - ar nėra nelogiškų (klaidingų) duomenų
'''

# Išvados ir rezultatai
- Apsirašėme mažiausiai 40 parametrų įtakojančių būsto kainą. Kai kuriems iš jų atskirai rankiniu būdu pasidarėme modelius. Palyginę su turimais duomenimis supratome, kad trūksta daug parametrų, norint išgauti gerą tikslumą.
- Duomenyse net labai svarbių parametrų trūksta reikšmių (pvz. yra NaN).

Tobulinimui:    
- Boxplot'ines paklaidas padaryti!

Sudėtingumas:    
- Pasiimta problema turi labai daug kintamųjų kurie įtakoja kainą. Būstas yra sudėtingas gaminys.
- 58 stulpeliai duomenyse yra labai daug per labai trumpą laiką. Kuo daugiau stulpelių tuo ilgesnis feature'ų įvertinimas (feature engineering'as).
- Nežinoma ispanų kalba, reikšmingai prailgina kiek laiko reikia skirti uždavinio sprendimui.
- apsunkintas grupavimas kintamųjų dėl daug kintamųjų (netelpa į vieną ekraną)
- pandas_profiling sutaupo laiko išvengiant programavimo nuo nulio savo funkcijų.

- Išanalizavus duomenis paaiškėjo, kad duomenų kokybė prasta:
  - Nėra stulpelio `portal` duomenų, tad neaišku iš kokių portalų imti duomenys.
  - trūksta stulpelių aprašymų
  - Nėra `latitude`, `longitude` stulpelių duomenų, o ten kur adresas gali būti įrašytas, jis įrašytas tik ~ 50 % atvejų, ir iš to reikėtų išsitraukti su Google Maps API GPS koordinates, bet tam reikia daugiau laiko. Nespėta gauti API raktą.
  - Iš 22 skaičių tipo stulepių, tik 12 turėjo duomenis, iš 12 turinčių duomenis **tik 2 stulpeliai turėjo pilnus gerus duomenis** (`buy_price`, `buy_price_by_area`), likę 8 stulpeliai turėjo NaN reikšmių tarp 0,6–93 %, `n_rooms` turėjo reikšmių su 0, o `rent_price` turėjo neigiamų reikšmių pvz. -34_590_276 (kai maksimali reikšmė 2517).
  - Iš 5 bool tipo stulpelių 2 (40 %) stulpeliai neduoda jokios naudos t. y. `is_rent_price_known` visada yra `False` (nors iš pačio stulpelio `rent_price` aišku, be to jame yra stipriai neteisingų/nelogiškų neigiamų reikšmių), o `is_buy_price_known` visada yra `True` (nors tą galime matyti ir iš stulpelio `buy_price`).
  - Duomenų tipo object stulpeliams:
    - Stulpelis `operation` neturi prasmės, nes visos reikšmės yra `sale`.
    - object tipo stulepliuose NaN reikšmių yra 1,8–93 %.
    - 20 object stulpelių yra realiai bool reikšmės, su galimybe turėti NaN reikšmę.
    - Esant trūkumui stulpelių aprašymui, neaišku ar bool stulpeliai turintys tik True ir NaN reikšmes, yra realiai visada True ir False, ar kai kurias atvejais taip, o kai kuriais atvejais (pvz. priklausomai nuo portalo) NaN reiškia, kad trūksta duomenų.
    - `raw_address` NaN(25 %) = `street_name` NaN(27 %) + `street_number` NaN(71 %)